src/main/java/com/reshiftsecurity/plugins/intellij/core/SecurityReportService.java

```diff
@@ -0,0 +1,176 @@
+import com.reshiftsecurity.results.SecurityIssue;
+import edu.umd.cs.findbugs.BugCollection;
+import edu.umd.cs.findbugs.BugInstance;
+import org.apache.commons.collections.CollectionUtils;
+import org.eclipse.jgit.lib.ConfigConstants;
+import org.eclipse.jgit.lib.Repository;
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.util.HashSet;
+import java.util.Set;
+
+@State(
+        name = "ReshiftSecurity-Intellij-Report",
+        storages = {
+                @Storage("$PROJECT_CONFIG_DIR$/reshiftsecurity-intellij-report.xml")
+        }
+)
+public class SecurityReportService implements PersistentStateComponent<SecurityReportService> {
+
+    @Nullable
+    @Override
+    public SecurityReportService getState() {
+        return this;
+    }
+
+    @Override
+    public void loadState(final SecurityReportService state) { XmlSerializerUtil.copyBean(state, this); }
+
+    @NotNull
+    public static SecurityReportService getInstance(@NotNull final Project project) {
+        SecurityReportService service = ServiceManager.getService(project, SecurityReportService.class);
+        service.projectName = project.getName();
+        try {
+            Repository existingRepo = new FileRepositoryBuilder()
+                    .setGitDir(new File(String.format("%s/.git", project.getBasePath())))
+                    .build();
+            service.isGitProject = true;
+            Set<String> remoteNames = existingRepo.getRemoteNames();
+            service.gitRepo = existingRepo.getConfig().getString(
+                    ConfigConstants.CONFIG_KEY_REMOTE,
+                    CollectionUtils.isEmpty(remoteNames) ? "origin" : remoteNames.stream().findFirst().get(),
+                    ConfigConstants.CONFIG_KEY_URL);
+            service.gitBranch = existingRepo.getBranch();
+        } catch (IOException e) {
+            // might not be a git repo, fail silently as this is not a core function of the plugin
+            e.printStackTrace();
+        }
+        return service;
+    }
+
+    @Tag
+    private Boolean isGitProject = false;
+
+    @Tag
+    private String gitBranch = "";
+
+    @Tag
+    private String gitRepo = "";
+
+    @Tag
+    private String projectName;
+
+    @Tag
+    public int totalVulnerabilityCount = 0;
+
+    @Tag
+    public int newVulnerabilityCount = 0;
+
+    @Tag
+    public int totalFixCount = 0;
+
+    @Tag
+    public int newFixCount = 0;
+
+    @Tag(value = "securityIssues")
+    @XCollection(elementName = "securityIssue")
+    public Set<SecurityIssue> securityIssues;
+
+    private boolean issueExistsInSet(SecurityIssue issue, Set<SecurityIssue> issueSet) {
+        if (CollectionUtils.isEmpty(issueSet))
+            return false;
+
+        for (SecurityIssue i: issueSet) {
```

Comment:
If we're using hashes to identify issues, it would be more natural to use a HashMap instead of a Set here.
If you key a HashMap with the issue hash, then your lookup is average constant time instead of linear. This doesn't scale well with the number of issues.

If max number of issues expected is low, and number of times you call this is also low, you can disregard the suggestion :)

Reply:
fair point, the number of issues are not always predictable. So yes I will change it to a HashMap